run focus tests in parallel where possible

bparees opened the pull request.

bparees commented:
@smarterclayton @stevekuznetsov ptal.  yes this breaks the `test/extended/core.sh --ginkgo.focus=foo` execution in favor of `FOCUS=foo test/extended/core.sh` but `--ginkgo.focus` always struck me as confusing anyway since the real flag (if you're just running ginkgo) is `-focus`)

(Our Makefile, meanwhile, converts `FOCUS=foo make test-extended` into a `test/extended/core.sh --ginkgo.focus` invocation, which only makes the chain harder to understand as it is)

it also breaks the ability to pass random arguments to extended.test via `test/extended/core.sh arg1 arg2`, but given that the function name is `os::test::extended::focus` not `os::test::extended::run_with_user_args` I don't feel bad about that.

I think this could be refactored even further to get rid of the the ::focus function entirely (moving the logic into core.sh) but let me get some feedback on the general direction first.

bparees commented:
/test extended_builds

bparees commented:
/test extended_builds
/retest

stevekuznetsov commented:
LGTM. Whenever I break workflows here @bparees you and your scrum are the first to get bitten by it ... so I would do a survey to see how commonly people are using the flows you are breaking. Would be good to ask also @miminar and the registry guys.

You could for `--ginkgo.focus` breaking work around that by parsing args, maybe.

stevekuznetsov commented:
Or, we could approach this from the angle that the user needs to be explicit and allow for a `PARALLEL=true` to add in the parallelism, using this in the jobs.

bparees commented:
> You could for --ginkgo.focus breaking work around that by parsing args, maybe.

I could but I really don't want to.  I'd prefer to just move to a common "set FOCUS to define your focus" single path.  `--ginkgo.focus` has always struck me as a leaked abstraction.

> Or, we could approach this from the angle that the user needs to be explicit and allow for a PARALLEL=true to add in the parallelism, using this in the jobs.

yeah I did consider that approach as well, the biggest problem is that in order to run parallel we have to run via ginkgo (not invoking extended.test directly, as this was doing previously) and that gets us into all kinda of arg manipulations at which point the path I implemented seems more straightforward and then people also don't have to worry about setting/not setting that variable.

stevekuznetsov commented:
Seems reasonable. I would not suggest finding the ginkgo focus and manipulating it to pass it in... but what is the failure case for someone passing in `--ginkgo.focus` after this? Do they know the api changed under their feet? if we can do a grep over args to figure it out and print out a message that may be low cost high reward

bparees commented on test/extended/templates/helpers.go:
```diff
@@ -94,13 +94,14 @@ func EnsureTSB(tsbOC *exutil.CLI) (osbclient.Client, *exec.Cmd) {
-		// this is weird, but we have to ignore this error in case some of this already exists (race between tests and the like)
-		//o.Expect(err).NotTo(o.HaveOccurred())
+		// If template tests run in parallel this could be created twice, we don't really care.
+		fmt.Fprintf(g.GinkgoWriter, "Error creating TSB resources: %v \n", err)
```
@deads2k ptal.

deads2k replied:
How is this different than `e2e.Logf`?

deads2k replied:
> @deads2k ptal.

Yeah, ignoring errors during a race seems fine.

bparees replied:
pretty sure it's not, but we use all sorts of varieties throughout the extended tests, I copied this from another one.  standardization isn't gonna happen here :)

deads2k commented on test/extended/templates/helpers.go:
```diff
@@ -94,13 +94,14 @@ func EnsureTSB(tsbOC *exutil.CLI) (osbclient.Client, *exec.Cmd) {
 		_, _, err := tsbOC.AsAdmin().WithoutNamespace().Run("create", "namespace").Args(tsbNS).Outputs()
-		o.Expect(err).NotTo(o.HaveOccurred())
+		// If template tests run in parallel this could be created twice, we don't really care.
+		fmt.Fprintf(g.GinkgoWriter, "Error creating TSB namespace %s: %v \n", tsbNS, err)
```
Might want to check to see if there actually was an error.

bparees replied:
doh. yeah.

deads2k commented on test/extended/templates/helpers.go:
```diff
@@ -94,13 +94,20 @@ func EnsureTSB(tsbOC *exutil.CLI) (osbclient.Client, *exec.Cmd) {
-		o.Expect(err).NotTo(o.HaveOccurred())
+		// If template tests run in parallel this could be created twice, we don't really care.
+		if err != nil {
+			fmt.Fprintf(g.GinkgoWriter, "Error creating TSB namespace %s: %v \n", tsbNS, err)
```
@bparees `e2e.Logf` is used less than five lines above this.

bparees commented:
/retest

bparees commented:
/retest

stevekuznetsov commented:
Looks broken:

bparees commented:
sigh. thanks.

bparees commented:
> Seems reasonable. I would not suggest finding the ginkgo focus and manipulating it to pass it in... but what is the failure case for someone passing in --ginkgo.focus after this? Do they know the api 

currently it would be silently ignored (along w/ any other args they were passing) and all tests would be run instead.

> changed under their feet? if we can do a grep over args to figure it out and print out a message that may be low cost high reward

yeah i guess.  i was really hoping to just excise all the ginkgo.focus stuff rather than maintain some sort of deprecated/warning behavior, but you're correct that it would be more polite to let people know.

Note that the Makefile is still going to pass --ginkgo.focus to these scripts as things stand right now.  I need to see what other suite scripts need to be updated if any.

bparees commented:
/retest

bparees commented:
@stevekuznetsov ok i added better checking for people still passing --ginkgo.focus.  I think networking.sh is the one weird script that 

1) respects focuses (uses extended.test)
2) does not leverage setup.sh

So I added extra logic in that one as well, @knobunc should probably sign off on that.

Otherwise, ptal and feel free to merge in my absence if everything (included extended_builds) is passing.

bparees commented:
/test extended_builds

stevekuznetsov commented on test/extended/builds/run_policy.go:
```diff
@@ -145,7 +145,7 @@ var _ = g.Describe("[builds][Slow] using build configuration runPolicy", func()
 				if build.Status.Phase == buildapi.BuildPhaseRunning {
 					latency := lastCompletion.Sub(time.Now())
-					o.Expect(latency).To(o.BeNumerically("<", 10*time.Second), "next build should have started less than 10s after last completed build")
+					o.Expect(latency).To(o.BeNumerically("<", 15*time.Second), "next build should have started less than 15s after last completed build")
```
Intended?

bparees replied:
yeah, running in parallel seems to have thrown the timing off a bit by slowing things down.

stevekuznetsov commented on test/extended/networking.sh:
```diff
@@ -232,6 +232,12 @@ function run-extended-tests() {
 
+  # this ${FOCUS} value will override the $focus_regex in the same way that
+  # the --ginkgo.focus argument did previously.
+  if [[ ! -z ${FOCUS} ]]; then
```
```sh
if [[ -n "${FOCUS:-}" ]]; then
```

stevekuznetsov commented on test/extended/networking.sh:
```diff
@@ -283,6 +289,11 @@ esac
 TEST_EXTRA_ARGS="$@"
 
+if [[ "$@[@]" =~ "ginkgo.focus" ]]; then
+      echo "ERROR: --ginkgo.focus is no longer supported, use FOCUS=foo <suite.sh> instead."
```
Instead of `echo` + `exit` use `os::log::fatal`

stevekuznetsov commented on test/extended/setup.sh:
```diff
@@ -4,15 +4,28 @@
+		echo "ERROR: --ginkgo.focus is no longer supported, use FOCUS=foo <suite.sh> instead."
+		exit 1
+	fi
+	if [[ -n ${FOCUS} ]]; then
```
```sh
if [[ -n "${FOCUS:-}" ]]; then
```

bparees commented:
@stevekuznetsov comments addressed, thanks.  anything else?

stevekuznetsov commented on test/extended/setup.sh:
```diff
@@ -4,15 +4,28 @@
-		extended.test "$@"
-		exit $?
+	if [[ "$@[@]" =~ "ginkgo.focus" ]]; then
+		echo "ERROR: --ginkgo.focus is no longer supported, use FOCUS=foo <suite.sh> instead."
```
`os::log::fatal`

stevekuznetsov commented:
/retest
/lgtm
